lib/SIL/Linker.cpp

```diff
@@ -58,7 +58,13 @@ bool SILLinkerVisitor::maybeAddFunctionToWorklist(SILFunction *F) {
   if (!F->isExternalDeclaration())
     return false;
 
-  if (isLinkAll() || hasSharedVisibility(F->getLinkage()))
+  // Note: PublicNonABI function declarations deserialize as
+  // HiddenExternal, then become SharedExternal after the body
+  // has been deserialized. So try deserializing HiddenExternal
+  // too.
+  if (isLinkAll() ||
+      hasSharedVisibility(F->getLinkage()) ||
+      F->getLinkage() == SILLinkage::HiddenExternal)
```

Comment:
Two things:

1. Can't you do a check if this is public non ABI rather than doing this for all HiddenExternal functions?
2. Maybe split this up into separate predicates to make the flow clearer? I.e.:

```
// If we are supposed to link everything, just do it.
if (isLinkAll()) {
   ...
}

// Shared functions are emitted with link once odr linkage, so we want to import their bodies
// and re-emit them in the file. So if we have shared visibility, deserialize.
if (hasSharedVisibility(F->getLinkage())) {
   ...
}

// Finally, if we have a hidden external function with public non abi linkage, deserialize. This is because *snip* insert explanation here.
if (isPublicNonABI(F) && F->getLinkage() == SILLinkage::HiddenExternal) {
  ...
}

return false;
```